core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestPushdownLimitIntoWindow.java

```diff
@@ -40,36 +40,14 @@
     @Test
     public void testLimitAboveWindow()
     {
-        ResolvedFunction rowNumberFunction = tester().getMetadata().resolveFunction(QualifiedName.of("row_number"), fromTypes());
-        tester().assertThat(new PushdownLimitIntoWindow(tester().getMetadata()))
-                .on(p -> {
-                    Symbol a = p.symbol("a");
-                    Symbol rowNumberSymbol = p.symbol("row_number_1");
-                    OrderingScheme orderingScheme = new OrderingScheme(
-                            ImmutableList.of(a),
-                            ImmutableMap.of(a, SortOrder.ASC_NULLS_FIRST));
-                    return p.limit(
-                            3,
-                            p.window(
-                                    new WindowNode.Specification(ImmutableList.of(a), Optional.of(orderingScheme)),
-                                    ImmutableMap.of(rowNumberSymbol, newWindowNodeFunction(rowNumberFunction, a)),
-                                    p.values(a)));
-                })
-                .matches(
-                        limit(3, topNRanking(
-                                pattern -> pattern
-                                        .specification(
-                                                ImmutableList.of("a"),
-                                                ImmutableList.of("a"),
-                                                ImmutableMap.of("a", SortOrder.ASC_NULLS_FIRST))
-                                        .maxRankingPerPartition(3)
-                                        .partial(false), values("a"))));
+        assertLimitAboveWindow("row_number");
+        assertLimitAboveWindow("rank");
     }
 
     @Test
     public void testConvertToTopNRowNumber()
```

Comment:
should there be similar test for ranking?

Reply:
As far as I know, this conversion is corresponding to [this part](https://github.com/trinodb/trino/blob/625eec14a5a37fe04ebe59d56cd548295140e462/core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushdownLimitIntoWindow.java#L94-L96). 

https://github.com/trinodb/trino/blob/625eec14a5a37fe04ebe59d56cd548295140e462/core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushdownLimitIntoWindow.java#L94-L96

`rank` function does not seem to be supported yet. Can we let `rank` supported as well?

Reply:
No, this optimization does not extend to `rank` or `dense_rank`. The reason is that an unpartitioned topN row number is guaranteed to produce no more than N results, while `rank` and `dense_rank` can produce an arbitrary number of rows (for example if all of the incoming rows had the same value value).